usr.bin/mdo/mdo.c

```diff
@@ -19,52 +20,277 @@
+		"\n"
+		"Examples:\n"
+		"  mdo -u alice id\n"
+		"  mdo -u 1001 -g wheel -G staff,operator /bin/sh\n"
+		"  mdo -u bob -s @,+wheel,+operator /usr/bin/id\n"
+		"  mdo -E 1002 -R 1003 -U 1004 /bin/id\n"
+	);
+	exit(1);
 }
 
 int
 main(int argc, char **argv)
 {
-	struct passwd *pw;
+	struct passwd *pw = NULL;
 	const char *username = "root";
+	const char *primary_group = NULL;
+	const char *supp_groups_str = NULL;
+	const char *group_mod_str = NULL;
 	struct setcred wcred = SETCRED_INITIALIZER;
 	u_int setcred_flags = 0;
 	bool uidonly = false;
 	int ch;
 
-	while ((ch = getopt(argc, argv, "u:i")) != -1) {
+	gid_t gid = -1;
+	bool override_gid = false;
+
+	bool supp_reset = false;
+	gid_t *supp_add = NULL, *supp_rem = NULL;
+	size_t add_count = 0, rem_count = 0;
+
+	const char *ruid_str = NULL, *svuid_str = NULL, *euid_str = NULL, *rgid_str = NULL, *svgid_str = NULL;
+
+	bool print_rule = false;
+	for (int i = 1; i < argc; i++) {
+		if (strcmp(argv[i], "--print-rule") == 0) {
+			print_rule = true;
+
+			for (int j = i; j < argc - 1; j++) {
+				argv[j] = argv[j+1];
+			}
+			argc--;
+			break;
+		}
+	}
+
+	while ((ch = getopt(argc, argv, "u:ig:G:s:U:R:E:P:Q:h")) != -1) {
 		switch (ch) {
 		case 'u':
 			username = optarg;
 			break;
 		case 'i':
 			uidonly = true;
 			break;
+		case 'g':
+			primary_group = optarg;
+			break;
+		case 'G':
+			supp_groups_str = optarg;
+			break;
+		case 's':
+			group_mod_str = optarg;
+			break;
+		case 'U':
+			ruid_str = optarg;
+			break;
+		case 'R':
+			svuid_str = optarg;
+			break;
+		case 'E':
+			euid_str = optarg;
+			break;
+		case 'P':
+			rgid_str = optarg;
+			break;
+		case 'Q':
+			svgid_str = optarg;
+			break;
+		case 'h':
+			usage();
 		default:
 			usage();
 		}
 	}
+
+	if (uidonly && (primary_group || supp_groups_str || group_mod_str))
+		errx(EXIT_FAILURE, "-i cannot be used with -g, -G, or -s");
+
 	argc -= optind;
 	argv += optind;
 
-	if ((pw = getpwnam(username)) == NULL) {
-		if (strspn(username, "0123456789") == strlen(username)) {
-			const char *errp = NULL;
-			uid_t uid = strtonum(username, 0, UID_MAX, &errp);
-			if (errp != NULL)
-				err(EXIT_FAILURE, "invalid user ID '%s'",
-				    username);
-			pw = getpwuid(uid);
+	if (username) {
+		if ((pw = getpwnam(username)) == NULL) {
+			if (strspn(username, "0123456789") == strlen(username)) {
+				const char *errp = NULL;
+				uid_t uid = strtonum(username, 0, UID_MAX, &errp);
+				if (errp != NULL)
+					err(EXIT_FAILURE, "invalid user ID '%s'", username);
+				pw = getpwuid(uid);
+			}
+			if (pw == NULL)
+				err(EXIT_FAILURE, "invalid username '%s'", username);
 		}
-		if (pw == NULL)
-			err(EXIT_FAILURE, "invalid username '%s'", username);
 	}
 
 	wcred.sc_uid = wcred.sc_ruid = wcred.sc_svuid = pw->pw_uid;
 	setcred_flags |= SETCREDF_UID | SETCREDF_RUID | SETCREDF_SVUID;
 
-	if (!uidonly) {
+	if (ruid_str) {
+		const char *errp = NULL;
+		wcred.sc_ruid = strtonum(ruid_str, 0, UID_MAX, &errp);
```

Comment:
We should apply here the same logic as for `-u`, i.e., accept usernames as well as plain IDs.  Please factor out that code into a separate function (see, e.g., functions `a_uid()` and `a_gid()` in `usr.sbin/chown/chown.c`).